### DISCUSSION TITLE
Improving review habits

### DISCUSSION THREAD

**Discussion by adeak** [pyvista/pyvista#2245] [Ideas]
As you all know one of my greater pet peeves is doing reviews right. As I also try to say each time this comes up, I don't expect anyone to be as pedantic (or meticulous if we're aiming for a more cheritable angle) as I am so that's not what I want to discuss.

There's been an increasing tendency of merging PRs which weren't quite properly reviewed. It's probably just a statistical fluke that these have happened more often, but it's enough for me to want to address it.

Reviewer time is expensive and most of us (myself especially included) can rarely invest the time to give PRs a thorough review. But there are some pain points that could (should) still be improved:
1. PRs ideally should never be merged without review. I can always imagine exceptions, like reverting changes that break downstream completely, or some security-related hotfix where the risk of getting the PR merged unreviewed is outweighed by the risk of inaction. And if longer time has passed where nobody was available to review then obviously merging can be reasonable in order to get stuff done. But ideally I'd still ask for a developer team ping a few days ahead as a heads-up that merging will commence even without reviews. Often people don't have time "right now" but when the other option is un unreviewed merge then contributors can try rearranging their workloads.
2. PRs shouldn't be merged in a rush without good reason, even if there's an approving review. I occasionally see non-trivial PRs opened and merged in a less-than-12-hour window, which is not great because contributors are scattered all over the globe and we should give a chance for everyone to take a look if they want to. We know from experience that different people often spot different kinds of issues.
3. PRs shouldn't be merged (time permitting, as always) without a final round of approvals on every change. What I mean is that if person A (not necessarily the PR author) pushes some change then at least one person (other than A) should look at those changes before merging, even if the changes before those commits were approved by others. This point includes the case of responding to review comments. There's usually enough time to wait for reviewers to respond to replies to review comments. A PR author might think that they addressed a review comment, but people often misunderstand each other, in which case letting the reviewer clarify or elaborate can help. Replies, thumbsup reactions and setting threads to "resolved" are all straightforward ways to signal approval from a reviewer.

The way I see it, the point of code review is the premise that people make mistakes. This applies to every change. Getting another pair of eyes on every piece of the _final_ state of a PR ensures that we catch some of those mistakes before they grow up to be bugs. Obviously we must weigh the cost of this process against the cost of "just getting things done". I'm not suggesting any protocol or formal constraint, just ask that we all use our best judgement and try to make sure that things are done right before pressing the big merge button.

(Just to explain my motivation: most of the mistakes that slip through are minor. They don't matter too much if they are merged. But there are plenty of papercut-sized issues already in the codebase, which nobody will ever get around to fix entirely. So the way I see it is that the best strategy is to at least ensure that we introduce as few new mistakes as we can to keep the problem sustainable. Easier to fix small issues at the start than finding them again years later and going through the trouble of fixing them.)

**Comment by tkoyama010** [comment]
@adeak Thank you for pointing up this topic. I think maintainers tend to be aggressive in merging because they are passionate about adding new fixes quickly and want to see the released library as soon as possible.
I agree with you that we should not make this a formal constraint as you say. However, No.3 was an important behavioral criterion that I had overlooked. I would like to add this to CONTRIBUTING.rst, how about it?

**Reply by adeak** [reply]
Thanks for your thoughts @tkoyama010.

I personally would consider adding it to CONTRIBUTING.rst "a formal constraint" :) And I don't think we have review-related things there, as most of the document is aimed at regular (and especially new) contributors, which makes sense. Maintainers probably rarely look at the document anyway, and in any case I don't imagine anyone would need to point a maintainer at the document. I mean if I or someone else would have a similar issue in the future, we would only have to ask the maintainers to be more considerate, but saying "as it is written in CONTRIBUTING.rst" would be overkill :)

**Comment by banesullivan** [comment]
Thanks for raising this, @adeak! I must admit that I'm often a bit trigger happy and merge premature PRs in the name of moving quick/"getting things done" and reducing the cost of reviews on progress. 

I think you raise some excellent criteria that we should strive for, though with our regular reviewers being a rather small team and all of those folks are graciously volunteering their time, we won't always be able achieve perfection. 

But, I agree, we can do better. I will commit to holding myself to a better standard when it comes to reviewing PRs to mitigate new slip-ups in the code base and to foster a better culture that encourages there to be more time for everyone to have a chance to review changes.

If we continue as is, we will never get to a v1.

**Comment by banesullivan** [comment]
One thing to keep in mind is that we follow trunk based development for PyVista's branching model: https://trunkbaseddevelopment.com/

This mode of branching/development encourages fast-paced contributions and mitigates the risk associated with introducing new minor issues as patch releases are actually patches and we can rely on any minor version of PyVista to remain stable.

To me, this mode of development encourages merging premature PRs with followup PRs to fix the small mistakes in the name of progress

Some notes on trunk based development: https://trunkbaseddevelopment.com/short-lived-feature-branches/

> One key rule is the length of life of the branch before it gets merged and deleted. Simply put, the branch should only last a couple of days. Any longer than two days, and there is a risk of the branch becoming a long-lived feature branch (the antithesis of trunk-based development).

> Working towards a pull request can sometimes make you look past the power of streaming a series of small commits into trunk for the benefit of all.

Perhaps we can take some of the best parts of trunk based development and your points on needing a thorough review process that can make developing PyVista a fun, inclusive, and rewarding experience for all involved.

**Reply by adeak** [reply]
Thanks @banesullivan. I guess my point is more of a technical one: it's a lot harder to fix slip-ups in follow-up PRs because the diff is only visible on the old PR, and one would have to hunt down each snippet to be fixed and fix what needs fixing there. It's just a lot more overhead than fixing them in the first place. And even if someone raises some points in comments on merged PRs preparing for a follow-up, often there's no reply. So the still open PR is the window of opportunity.

The way I see it is that the "branches live for a few days" mentioned at trunk-based development is aimed for the extreme case of a library being developed with lots of commits to main per day. In that case taking your time means gradually increasing the risk of conflicts which is to be avoided. But when there's no huge activity on main, nor an impending release, and things get merged for merging's sake, that's what I see as suboptimal :)

**Reply by banesullivan** [reply]
I think you're spot on here

Development for PyVista isn't as fast-paced and we have a small team of contributors, so we can definitely work to improve our review process to mitigate mistakes.

**Reply by adam-grant-hendry** [reply]
> Any longer than two days, and there is a risk of the branch becoming a long-lived feature branch (the antithesis of trunk-based development).

If we ask developers to `git pull --rebase` their feature branches regularly, would this overcome that problem? The branch might be "long-lived" from a time aspect, but it would be up to date with the business logic and minimize merge conflicts. (I still think merges should have as soon as possible though, don't get me wrong).

Limiting branch life to an arbitrary time length, IMO, is aimed at minimizing merge conflicts and a feature getting stale (someone forgetting what they were doing). Bug-wise, I think the bigger factor is making many small commits as compared to fewer large commits so you reduce the potential for creating a big breaking change that's hard to revert.

**Reply by adeak** [reply]
> If we ask developers to `git pull --rebase` their feature branches regularly, would this overcome that problem?

I don't think so. We rarely get any conflicts because most of our PRs are small and like I said, development itself is slow-paced. The few conflicts are usually trivial to handle. More importantly, if there _were_ larger, accumulated conflicts, it would be due to contributors not being able to spend more time on the PR in the given time frame. Telling them to rebase often wouldn't solve this: if they could rebase then they would probably already do that.

**Comment by banesullivan** [comment]
FYI, I have increased the number of required reviews to 2 - I'm hoping that this will prevent auto-merges from firing after 1 good review

**Reply by adeak** [reply]
Thanks for the heads-up. I hope this won't stall PRs, because that's not my intention at all.

**Reply by tkoyama010** [reply]
Thank you for increasing the number of required reviewers. The number of community members is growing, so I think it's possible to have two people reviewing on a regular basis.

**Reply by adeak** [reply]
We're back to 1 approval for merge, right? Cc @akaszynski

**Reply by akaszynski** [reply]
Correct. Mentioned this in the slack as well.

Be sparing with your reviews if you're not sure about complex PRs, especially on the longer ones.

**Reply by adeak** [reply]
I don't see it anywhere in slack @akaszynski. I only saw a PR being merged with a single approval... This needs visibility/transparency.

**Reply by akaszynski** [reply]
My bad. Sometimes I assume (because I'm not on slack that often) that enough people have seen it.

I agree with your comments earlier, I just found it a bit much with two approvers, especially as Bane's been less active as of late.

If this doesn't work out with one, we can go back to two.

**Reply by adeak** [reply]
It's all good going back to 1, it's just that devs should know. And when I say "I don't see it anywhere in slack" I mean it, I still don't see it, even if I go looking for it :D

**Reply by akaszynski** [reply]
I had to hunt for it as well:
https://pyvista.slack.com/archives/CQ2L29K63/p1650291821182939?thread_ts=1646320245.624919&cid=CQ2L29K63

**Reply by adeak** [reply]
I see, thanks @akaszynski. That's a thread from 2 months ago, and the only people who are notified of new messages are people involved in the thread. So Bane and yourself in this case.

So heads-up @pyvista/developers: now we're back to 1 approval needed for merging, please account for that. I'll post a message in the slack channel too, just in case.

**Reply by akaszynski** [reply]
Turns out I'm bad at slack...

**Comment by tkoyama010** [comment]
I would like to refrain from merging for at least 24 hours after I submit my LGTM. During that time the earth will turn a full revolution and all interested reviewers will be given the opportunity to review. However, I would prefer to automatically merge after 24 hours if there are no comments. I will have to look at the same PR again after 24 hours without comments. The current enable auto marge does not have the option to specify a time, I wish GitHub Action could do that.

**Reply by adeak** [reply]
Good idea, it would be great to have an auto-merge with a 1-day timer. Lift the burden from having to check back, but also allowing eyes around the globe to take a look.

**Reply by akaszynski** [reply]
That's a great idea!

One way to implement that would be to create a github action cronjob that runs nightly and checks if there are any PRs that have been approved without additional comments after the initial approval. I'm thinking of using [pygithub](https://github.com/PyGithub/PyGithub). We can add it to the "to-do" list unless someone has a better approach

**Reply by tkoyama010** [reply]
Thank you for your comments. It was great to know that everyone's needs are the same. However, we need to do some research on how to make it.

**Reply by tkoyama010** [reply]
I solve this in https://github.com/pyvista/pyvista/pull/2851 .
Also @akaszynski could you please add the following additional settings to the master branch?
![TO2HN](https://user-images.githubusercontent.com/7513610/175551826-586cb2cf-ce5a-4714-b8e4-1ee722dba566.png)
I was concerned that after I Approval it would add sources that I have not reviewed. I do not currently have permission to do this.

**Reply by banesullivan** [reply]
Done

**Reply by akaszynski** [reply]
Agree with this. Thanks @banesullivan.

**Reply by tkoyama010** [reply]
Thanks

**Comment by adam-grant-hendry** [comment]
@adeak I love the enthusiasm! I am a stickler for detail, so I appreciate wanting to review more. As a new contributor, I depend on the eyes of the veterans to help me in case anything I've done would cause problems. I like the 24-hr waiting period idea by @tkoyama010 and the 2-person review by @banesullivan .

Though I'm loathe of GitLab over GitHub, I use it at work and it has one interesting feature: Merge Request Pipelines. The feature is currently half-baked in GitLab since you must use merge commits and cannot squash merge and fast-forward merge, though they aim to fix that. However, it creates a private branch that runs tests against the merged code prior to merging to see if any problems would occur.

Does anyone know if `GitHub Actions` supports pre-merge checks? (I'm not sure what bugs have slipped through though).

Another option, though perhaps not entirely popular, would be to switch from trunk-based development to GitHub flow development. The only change would be using merge commits and not squashing. The benefit would be one could go back into a previous feature branch and run `git bisect` to identify the offending commit that introduced a bug.

Does anyone think using `GitHub Flow` in the future would help?

**Reply by adeak** [reply]
> it creates a private branch that runs tests against the merged code prior to merging to see if any problems would occur.

Isn't that exactly how CI works anyway?

> Another option, though perhaps not entirely popular, would be to switch from trunk-based development to GitHub flow development. 

I wouldn't want to change anything like this, the current model is fine.

More importantly, no amount of automatic checks will help. We already have an extensive test suite. I really just want for people to have the option to look at changes if they want to, and for no change to go in unreviewed (or barely reviewed). For instance, even though we build the docs in CI, someone has to go look at the build artifact to spot markup issues that don't manifest as errors during the build.

**Reply by adam-grant-hendry** [reply]
> Isn't that exactly how CI works anyway?

Not with a 3-way merge, at least in GitLab. It merges your code first, then runs checks. Your checks can fail, in which case you have to commit a revert...perhaps this is just a GitLab thing.

> I wouldn't want to change anything like this, the current model is fine....I really just want for people to have the option to look at changes if they want to, and for no change to go in unreviewed (or barely reviewed)

Totally agree, no complaints here.

I had another thought: Would it be worthwhile to use something like [Zoom/Meetups](https://github.com/pyvista/pyvista/discussions/2662) to do peer programming? I imagine peer programming is hard for open-source projects, but thought I'd ask.

**Reply by tkoyama010** [reply]
I have seen this done before with Sprint on Discord. We do not believe it is common in other situations.

**Reply by banesullivan** [reply]
> Would it be worthwhile to use something like https://github.com/pyvista/pyvista/discussions/2662 to do peer programming? I imagine peer programming is hard for open-source projects, but thought I'd ask.

I'd love to see this happen and I'd love to participate though I don't know if I can currently be the one to organize these
